qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py

```diff
@@ -0,0 +1,351 @@
+
+
+def _is_antisymmetric(mat: np.ndarray, rtol: float = 1e-5, atol: float = 1e-8) -> bool:
+    return np.allclose(mat, -mat.T, rtol=rtol, atol=atol)
+
+
+class QuadraticHamiltonian(TolerancesMixin):
+    r"""A Hamiltonian that is quadratic in the fermionic ladder operators.
+
+    A quadratic Hamiltonian is an operator of the form
+
+    .. math::
+        \sum_{p, q} M_{pq} a^\dagger_p a_q
+        + \frac12 \sum_{p, q}
+            (\Delta_{pq} a^\dagger_p a^\dagger_q + \text{h.c.})
+        + \text{constant}
+
+    where :math:`M` is a Hermitian matrix and :math:`\Delta` is an antisymmetric matrix.
+    """
+
+    def __init__(
+        self,
+        hermitian_part: Optional[np.ndarray] = None,
+        antisymmetric_part: Optional[np.ndarray] = None,
+        constant: float = 0.0,
+        num_modes: Optional[int] = None,
+        validate: bool = True,
+        rtol: float = 1e-5,
+        atol: float = 1e-8,
+    ) -> None:
+        r"""Initialize a QuadraticHamiltonian.
+
+        Args:
+            hermitian_part: The matrix :math:`M` containing the coefficients of the terms
+                that conserve particle number.
+            antisymmetric_part: The matrix :math:`\Delta` containing the coefficients of
+                the terms that do not conserve particle number.
+            constant: An additive constant term.
+            num_modes: Number of fermionic modes. This should be consistent with hermitian_part
+                and antisymmetric_part if they are specified.
+            validate: Whether to validate the inputs.
+            rtol: Relative numerical tolerance for input validation.
+            atol: Absolute numerical tolerance for input validation.
+
+        Raises:
+            ValueError: Either Hermitian part, antisymmetric part, or number of modes must
+                be specified.
+            ValueError: Hermitian part and antisymmetric part must have same shape.
+            ValueError: Hermitian part must have shape num_modes x num_modes.
+            ValueError: Hermitian part must be Hermitian.
+            ValueError: Antisymmetric part must have shape num_modes x num_modes.
+            ValueError: Antisymmetric part must be antisymmetric.
+        """
+        if num_modes is not None:
+            self._num_modes = num_modes
+        elif hermitian_part is not None:
+            self._num_modes, _ = hermitian_part.shape
+        elif antisymmetric_part is not None:
+            self._num_modes, _ = antisymmetric_part.shape
+        else:
+            raise ValueError(
+                "Either Hermitian part, antisymmetric part, or number of modes must be specified."
+            )
+
+        if validate:
+            if (
+                hermitian_part is not None
+                and antisymmetric_part is not None
+                and hermitian_part.shape != antisymmetric_part.shape
+            ):
+                raise ValueError("Hermitian part and antisymmetric part must have same shape.")
```

Comment:
In these error messages it is helpful for a user and debugging etc for them to be formatted with the actual failing values.

Reply:
Done.